src/arrayMethodReduce.js

```diff
@@ -5,7 +5,37 @@
  */
 function applyCustomReduce() {
   [].__proto__.reduce2 = function(callback, initialValue) {
-    // write code here
+    if (this === null) {
+      throw new Error('Cant iterate over underfined or null');
+    }
+
+    let result;
+    let i = 0;
+
+    const obj = Object(this);
+    const len = obj.length;
+
+    if (typeof callback !== 'function') {
+      throw new Error('callback is not a function');
+    }
+
+    if (arguments.length >= 2) {
+      result = initialValue;
+    } else {
+      if (len === 0) {
+        throw new Error('reduce of empty array with no initial value');
+      }
+      result = obj[i];
```

Comment:
you don't need all these custom errors here
simplify your solution
js will throw an error where it will be needed